lib/mapping/MapEditUtils.cpp

```diff
@@ -0,0 +1,371 @@
+
+const std::string TerrainViewPattern::RULE_DIRT = "D";
+const std::string TerrainViewPattern::RULE_SAND = "S";
+const std::string TerrainViewPattern::RULE_TRANSITION = "T";
+const std::string TerrainViewPattern::RULE_NATIVE = "N";
+const std::string TerrainViewPattern::RULE_NATIVE_STRONG = "N!";
+const std::string TerrainViewPattern::RULE_ANY = "?";
+
+TerrainViewPattern::TerrainViewPattern() : diffImages(false), rotationTypesCount(0), minPoints(0)
+{
+	maxPoints = std::numeric_limits<int>::max();
+}
+
+TerrainViewPattern::WeightedRule::WeightedRule(std::string& Name) : points(0), name(Name)
+{
+	standardRule = (TerrainViewPattern::RULE_ANY == Name || TerrainViewPattern::RULE_DIRT == Name
+		|| TerrainViewPattern::RULE_NATIVE == Name || TerrainViewPattern::RULE_SAND == Name
+		|| TerrainViewPattern::RULE_TRANSITION == Name || TerrainViewPattern::RULE_NATIVE_STRONG == Name);
+	anyRule = (Name == TerrainViewPattern::RULE_ANY);
+	dirtRule = (Name == TerrainViewPattern::RULE_DIRT);
+	sandRule = (Name == TerrainViewPattern::RULE_SAND);
+	transitionRule = (Name == TerrainViewPattern::RULE_TRANSITION);
+	nativeStrongRule = (Name == TerrainViewPattern::RULE_NATIVE_STRONG);
+	nativeRule = (Name == TerrainViewPattern::RULE_NATIVE);
+}
+
+void TerrainViewPattern::WeightedRule::setNative()
+{
+	nativeRule = true;
+	standardRule = true;
+	//TODO: would look better as a bitfield
+	dirtRule = sandRule = transitionRule = nativeStrongRule = anyRule = false; //no idea what they mean, but look mutually exclusive
+}
+
+CTerrainViewPatternConfig::CTerrainViewPatternConfig()
+{
+	const JsonNode config(ResourceID("config/terrainViewPatterns.json"));
+	static const std::string patternTypes[] = { "terrainView", "terrainType" };
+	for (int i = 0; i < ARRAY_COUNT(patternTypes); ++i)
+	{
+		const auto& patternsVec = config[patternTypes[i]].Vector();
+		for (const auto& ptrnNode : patternsVec)
+		{
+			TerrainViewPattern pattern;
+
+			// Read pattern data
+			const JsonVector& data = ptrnNode["data"].Vector();
+			assert(data.size() == 9);
+			for (int j = 0; j < data.size(); ++j)
+			{
+				std::string cell = data[j].String();
+				boost::algorithm::erase_all(cell, " ");
+				std::vector<std::string> rules;
+				boost::split(rules, cell, boost::is_any_of(","));
+				for (std::string ruleStr : rules)
+				{
+					std::vector<std::string> ruleParts;
+					boost::split(ruleParts, ruleStr, boost::is_any_of("-"));
+					TerrainViewPattern::WeightedRule rule(ruleParts[0]);
+					assert(!rule.name.empty());
+					if (ruleParts.size() > 1)
+					{
+						rule.points = boost::lexical_cast<int>(ruleParts[1]);
+					}
+					pattern.data[j].push_back(rule);
+				}
+			}
+
+			// Read various properties
+			pattern.id = ptrnNode["id"].String();
+			assert(!pattern.id.empty());
+			pattern.minPoints = static_cast<int>(ptrnNode["minPoints"].Float());
+			pattern.maxPoints = static_cast<int>(ptrnNode["maxPoints"].Float());
```

Comment:
isn't it possible to read `int` directly from json?

Reply:
this code was just moved from one place to another as I understand. Refactoring is not a part of current pr